### ISSUE TITLE
Wishlist: Pkg.add option to ask for confirmation and report changes

### ISSUE THREAD

**Issue by mashu** [JuliaLang/Pkg.jl#1945]
Hi,

It would be nice if Pkg.add had an option that informs user what it is going to change and requiring confirmation.
It does not need to be default, but at least some sorts of report before accepting any changes to current environment.

Thanks,

**Comment by KristofferC** [comment]
You can just do `undo` after a package operation.

**Comment by mashu** [comment]
> You can just do `undo` after a package operation.

Thanks, that's useful. Does it revert back also pre-compiled binaries? It takes a while to build those.
But please consider adding such an option to warn user about changes. 

Thanks!

**Comment by StefanKarpinski** [comment]
Unclear what you're referring to. Are you talking about prebuilt binary dependencies that are downloaded? Or precompile files ie `.ji` files?

**Comment by mashu** [comment]
> Unclear what you're referring to. Are you talking about prebuilt binary dependencies that are downloaded? Or precompile files ie `.ji` files?

I am sorry I wasn't aware that binary dependencies that are downloaded can be compiled. I think I am referring to anything that is build by "precompile" command.

**Comment by StefanKarpinski** [comment]
Precompilation doesn't happen unless load the package or explicitly ask for it and the ten most recent `.ji` files for each package are kept around in case they are useful again, so there doesn't seem to be any change needed here.

**Comment by Vexatos** [comment]
As #1937 was declared a duplicate of this issue, I'll repeat my suggestion from there.
`Pkg.add`, `Pkg.rm`, and `Pkg.update` should all support this. There should be a non-interactive version that merely produces the changes to be done without actually applying any and without asking for user confirmation. Ideally, this could be in the form of a list or dict that is easy to parse using update automation programs.

**Comment by StefanKarpinski** [comment]
There are multiple functionalities that seem to me to make this unnecessary:

1. Updates are non-destructive: all an update does is install new versions which can be used instead; the old versions are all still there and usable still.
1. Changes are tracked in project-local project & manifest files, which can be version controlled and which you can therefore see changes to, e.g. by doing `git diff Project.toml Manifest.toml`; if you revert the changes to the last checked in state, you have undone everything perfectly, and because of the first point that state is guaranteed to still work.
2. There is already `pkg> undo` which undoes the last operation: if you didn't like it, then undo it. This works even if you happen to have done multiple changes without making a git commit to your project.

So what would the benefit of a preview option be? Note that we previously *had* a preview option, but it was not very helpful because it's quite hard to implement: you have to thread the option through *all* operations and they all have to work in a mode where they make changes as they go or don't.

Ultimately I'm somewhat sympathetic to this for a very different reason: I want Pkg to be rearchitected so that it makes a plan of what it's going to do without actually doing anything and then hands that plan off to a separate state that goes and does all the things (in parallel). In that architecture, this feature becomes easier because it is just a way of presenting the plan to the user without executing it. In that design every command operates in three stages:

1. Pre-command preparations (e.g. update registries)
2. Parse/interpret command
3. Plan how to execute command
4. Execute the plan

In this scheme, you could prefix any command with `preview` and it would change the last step to "Display the plan" instead of executing it. Once we have that design, we could have this relatively easily, but until then this does seem like a fairly superfluous feature given how Pkg already works.

**Comment by Vexatos** [comment]
I can only speak for my own use case as an example. What we have is essentially a wrapper around `Pkg` in our UI application in the form of a "Check for Updates" button which lets the user check for updates before performing any actual update. Actually performing the update can take an immense amount of time (as explained in #1937), especially on Windows and on older hardware (a significant number of our users use laptops older than seven years). Updating the registries alone can take multiple minutes, although Julia 1.7 may partially solve this.
Even if updates are non-destructive, they often take so long that many people have to specifically allocate time for it. Getting a preview would allow our program to check whether any updates are available at all and, if there are, would be able to tell how many packages would have to be updated, giving at least a rough estimate of the time the update would take. The way we currently check whether an update is available at all is by manually comparing the git hash found in the `.julia/clones` directory for out program (which is installed through git) with the latest hash found online, which is obviously far from ideal but the only solution available at this time. The old `preview=true` option wasn't ideal either but by rerouting stdout, one could at least parse the output text.
The main use case for a machine-readable (ideally in the form of a vector or dict, as mentioned) would be integration with UI frameworks targeted at a non-technical audience. Not everyone is able to simply ask their users to run a git command or open a Julia REPL, sadly. Our users will only ever update packages through our UI, and many have never opened a terminal in their life.

**Comment by mashu** [comment]
I second that opinion above, having seen a list of packages pending update
would give user rough estimate of time, it can take. Furthermore, some
packages might have artificat dependencies which would mean additional
download time, extreme case is Flux which seems to be downloading CUDA, and
take takes ~15min on my laptop. It's not Pkg fault, but in general
artifacts add to compile time.
There could be also a reason not to update if you think, it could break
some of the stuff you're working on currently and you might want to delay
the update until you finish your coding, but at the same time you need some
functionality and you need to at this very moment decide from which package
to get it that is actually not breaking your current project.
Finally, there is another system Gentoo distribution, which uses git for
their packages. As far as I remember it first presents plan before
updating/compiling packages, maybe it's a good place to see how they
implemented that.

**Comment by StefanKarpinski** [comment]
> It's not Pkg fault, but in general artifacts add to compile time.

Artifacts don't involve any compilation, for what it's worth. You just download and unpack them. There is no compilation step. We don't generally use git for package installation anymore, everything is downloaded from package servers first.

Anyway, having a way of presenting what would get updated if a command were run would be possible with the rearchitecting that I'd like to do. It will take a while to get there, but this feature can be added at that point. It might also be good to have a way of knowing how big all the packages and artifacts that are going to be installed are since that would be a consideration for people.
